presto-kafka/src/main/java/io/prestosql/plugin/kafka/encoder/avro/AvroRowEncoder.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.kafka.encoder.avro;
+
+import io.prestosql.plugin.kafka.encoder.EncoderColumnHandle;
+import io.prestosql.plugin.kafka.encoder.EncoderFieldValueProvider;
+import io.prestosql.plugin.kafka.encoder.RowEncoder;
+import io.prestosql.spi.PrestoException;
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileWriter;
+import org.apache.avro.generic.GenericData;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.avro.io.DatumWriter;
+
+import java.io.ByteArrayOutputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.prestosql.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static java.util.Objects.requireNonNull;
+import static java.util.function.Function.identity;
+
+public class AvroRowEncoder
+        implements RowEncoder
+{
+    public static final String NAME = "avro";
+    private final DatumWriter<GenericRecord> avroRecordWriter;
+    private final Schema schema;
+    private final Map<EncoderColumnHandle, AvroColumnEncoder> columnEncoders;
+
+    public AvroRowEncoder(DatumWriter<GenericRecord> avroRecordWriter, Schema schema, Set<EncoderColumnHandle> columns)
+    {
+        this.avroRecordWriter = requireNonNull(avroRecordWriter, "avroRecordWriter is null");
+        this.schema = requireNonNull(schema, "schema is null");
+        columnEncoders = columns.stream().collect(toImmutableMap(identity(), this::createColumnEncoder));
+    }
+
+    private AvroColumnEncoder createColumnEncoder(EncoderColumnHandle columnHandle)
+    {
+        return new AvroColumnEncoder(columnHandle);
+    }
+
+    @Override
+    public byte[] encodeRow(List<EncoderFieldValueProvider> rowData)
+    {
+        GenericData.Record record = new GenericData.Record(schema);
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        rowData.forEach(fieldValueProvider -> columnEncoders.get(fieldValueProvider.getColumn()).encodeField(fieldValueProvider, record));
+        try (DataFileWriter<GenericRecord> dataFileWriter = new DataFileWriter<>(avroRecordWriter)) {
+            dataFileWriter.create(schema, outputStream);
+            dataFileWriter.append(record);
```

Comment:
Perhaps we should consider changing the `RowEncoder` interface to return a `ProducerRecord` instead of a `byte[]`.  That would allow us to just create a `GenericData.Record` and let the Kafka producer worry about serialization.  Seems like a safer approach.  WDYT?

Reply:
Would that make the encoder kafka specific?

Reply:
It lives in the Kafka module, right :)  I know you envisioned this as a parallel to `RecordDecoder`, but it's not at this point and it shouldn't prevent us from making the right decision here.  I don't know if `RowEncoder` can be a connector-agnostic interface - I don't know for how many connectors a `byte[]` is sufficient for writing.  And we really wouldn't gain much by having this be connector-agnostic.